feat(service): upgrade trigger template to v4

VictorCano opened the pull request:
## Changes

- feat(service): upgrade trigger template to v4 with all the necessary changes

## Migration notes for v3 users

1. Export data from v3 PostgreSQL
2. Deploy fresh v4 stack (includes ClickHouse and MinIO)
3. Migrate data to v4 database schema
4. Update worker configurations for new supervisor architecture

Refer to [Trigger.dev v4 docker docs](https://trigger.dev/docs/self-hosting/docker) for more details.

VictorCano committed:
feat(templates): add Trigger.dev v4 self-hosted template

- Complete v4 architecture with supervisor and managed workers
- Docker socket proxy for secure container management
- Support for private container registries (GHCR)
- GitHub OAuth and email support (Resend + SMTP)
- ClickHouse for run replication and analytics
- MinIO for object storage with auto-bucket creation
- ElectricSQL 1.0.24 for real-time sync with healthcheck
- All services with health checks and proper dependencies
- Supervisor uses file:// token path for cleaner config
- Environment variables support Coolify UI override

VictorCano committed:
fix(trigger-v4): remove ELECTRIC_INSECURE flag incompatible with ELECTRIC_SECRET

Electric v1.0.24 validates that ELECTRIC_SECRET and ELECTRIC_INSECURE=true
cannot be set simultaneously. Since we're using ELECTRIC_SECRET for proper
security, remove the ELECTRIC_INSECURE flag.

VictorCano committed:
fix(trigger-v4): add defaults for ENCRYPTION_KEY and DEPLOY_REGISTRY_NAMESPACE

- ENCRYPTION_KEY must be exactly 32 bytes, provide default
- DEPLOY_REGISTRY_NAMESPACE is required, default to 'trigger'

VictorCano committed:
fix(templates): change Trigger.dev default email transport to log

Resend email transport requires RESEND_API_KEY, causing the webapp to crash
when no API key is provided. Default to 'log' transport which doesn't require
configuration and simply logs emails to console.

VictorCano committed:
fix(trigger-v4): use SERVICE_PASSWORD_64 for secrets and smtp for email

- Change SESSION_SECRET and MAGIC_LINK_SECRET from SERVICE_PASSWORD_32_*
  to SERVICE_PASSWORD_64_* because Coolify's 32-character password magic
  variables return empty values (bug), while 64-character variants work
- Change EMAIL_TRANSPORT default from 'log' to 'smtp' because 'log' is
  not a valid option - valid values are: resend, smtp, aws-ses

This fixes the Trigger.dev container crash loop caused by:
- Missing MAGIC_LINK_SECRET and SESSION_SECRET env variables
- Invalid EMAIL_TRANSPORT enum value

All 8 services now start and become healthy:
- trigger, supervisor, docker-proxy, electric (applications)
- clickhouse, minio (applications)
- postgresql, redis (databases)

ShadowArcanist commented on templates/compose/trigger-v4.yaml:
```diff
@@ -0,0 +1,275 @@
+  clickhouse-data:
+  clickhouse-logs:
+  minio-data:
+  shared-data:
```
```suggestion
```
These are not needed (Coolify adds them automatically)

ShadowArcanist commented on templates/compose/trigger-v4.yaml:
```diff
@@ -0,0 +1,275 @@
+      start_period: 30s
+
+  minio:
+    image: minio/minio:latest
```
```suggestion
    image: ghcr.io/coollabsio/minio:RELEASE.2025-10-15T17-29-55Z
```
minio/minio:latest is no longer updated and the last released version is vulnerable to CVE: https://github.com/minio/minio/security/advisories/GHSA-jjjj-jwhf-8rgr

ShadowArcanist commented on templates/compose/trigger-v4.yaml:
```diff
@@ -0,0 +1,275 @@
+
+  clickhouse:
+    image: clickhouse/clickhouse-server:latest
+    restart: unless-stopped
```
```suggestion
```
These are not needed (Coolify adds them automatically)

ShadowArcanist commented on templates/compose/trigger-v4.yaml:
```diff
@@ -0,0 +1,275 @@
+
+  electric:
+    image: electricsql/electric:1.0.24
+    restart: unless-stopped
```
```suggestion
```
These are not needed (Coolify adds them automatically)

ShadowArcanist commented on templates/compose/trigger-v4.yaml:
```diff
@@ -0,0 +1,275 @@
+
+  redis:
+    image: redis:7
+    restart: unless-stopped
```
```suggestion
```
These are not needed (Coolify adds them automatically)

ShadowArcanist commented on templates/compose/trigger-v4.yaml:
```diff
@@ -0,0 +1,275 @@
+
+  postgresql:
+    image: postgres:16-alpine
+    restart: unless-stopped
```
```suggestion
```
These are not needed (Coolify adds them automatically)

ShadowArcanist commented on templates/compose/trigger-v4.yaml:
```diff
@@ -0,0 +1,275 @@
+
+  docker-proxy:
+    image: tecnativa/docker-socket-proxy:latest
+    restart: unless-stopped
```
```suggestion
```
These are not needed (Coolify adds them automatically)

ShadowArcanist commented on templates/compose/trigger-v4.yaml:
```diff
@@ -0,0 +1,275 @@
+
+  supervisor:
+    image: 'ghcr.io/triggerdotdev/supervisor:main'
+    restart: unless-stopped
```
```suggestion
```
These are not needed (Coolify adds them automatically)

ShadowArcanist commented on templates/compose/trigger-v4.yaml:
```diff
@@ -0,0 +1,275 @@
+services:
+  trigger:
+    image: 'ghcr.io/triggerdotdev/trigger.dev:v4-beta'
+    restart: unless-stopped
```
```suggestion
```
These are not needed (Coolify adds them automatically)

ShadowArcanist commented on templates/compose/trigger-v4.yaml:
```diff
@@ -0,0 +1,275 @@
+
+  minio:
+    image: minio/minio:latest
+    restart: unless-stopped
```
```suggestion
```
These are not needed (Coolify adds them automatically)

ShadowArcanist commented on templates/compose/trigger-v4.yaml:
```diff
@@ -0,0 +1,275 @@
+    image: 'ghcr.io/triggerdotdev/trigger.dev:v4-beta'
+    restart: unless-stopped
+    networks:
+      - trigger-net
```
```suggestion
```
One click services are just a compose based application so they are put into their own docker network by default, so adding network like this is not needed

ShadowArcanist commented on templates/compose/trigger-v4.yaml:
```diff
@@ -0,0 +1,275 @@
+    image: 'ghcr.io/triggerdotdev/supervisor:main'
+    restart: unless-stopped
+    networks:
+      - trigger-net
```
```suggestion
```
One click services are just a compose based application so they are put into their own docker network by default, so adding network like this is not needed

ShadowArcanist commented on templates/compose/trigger-v4.yaml:
```diff
@@ -0,0 +1,275 @@
+    image: tecnativa/docker-socket-proxy:latest
+    restart: unless-stopped
+    networks:
+      - trigger-net
```
```suggestion
```
One click services are just a compose based application so they are put into their own docker network by default, so adding network like this is not needed

ShadowArcanist commented on templates/compose/trigger-v4.yaml:
```diff
@@ -0,0 +1,275 @@
+    image: postgres:16-alpine
+    restart: unless-stopped
+    networks:
+      - trigger-net
```
```suggestion
```
One click services are just a compose based application so they are put into their own docker network by default, so adding network like this is not needed

ShadowArcanist commented on templates/compose/trigger-v4.yaml:
```diff
@@ -0,0 +1,275 @@
+    image: redis:7
+    restart: unless-stopped
+    networks:
+      - trigger-net
```
```suggestion
```
One click services are just a compose based application so they are put into their own docker network by default, so adding network like this is not needed

ShadowArcanist commented on templates/compose/trigger-v4.yaml:
```diff
@@ -0,0 +1,275 @@
+    image: electricsql/electric:1.0.24
+    restart: unless-stopped
+    networks:
+      - trigger-net
```
```suggestion
```
One click services are just a compose based application so they are put into their own docker network by default, so adding network like this is not needed

ShadowArcanist commented on templates/compose/trigger-v4.yaml:
```diff
@@ -0,0 +1,275 @@
+    image: clickhouse/clickhouse-server:latest
+    restart: unless-stopped
+    networks:
+      - trigger-net
```
```suggestion
```
One click services are just a compose based application so they are put into their own docker network by default, so adding network like this is not needed

ShadowArcanist commented on templates/compose/trigger-v4.yaml:
```diff
@@ -0,0 +1,275 @@
+    image: minio/minio:latest
+    restart: unless-stopped
+    networks:
+      - trigger-net
```
```suggestion
```
One click services are just a compose based application so they are put into their own docker network by default, so adding network like this is not needed

ShadowArcanist commented on templates/compose/trigger-v4.yaml:
```diff
@@ -0,0 +1,275 @@
+
+networks:
+  trigger-net:
+    driver: bridge
```
```suggestion
```
One click services are just a compose based application so they are put into their own docker network by default, so adding network like this is not needed

ShadowArcanist commented on templates/compose/trigger-v4.yaml:
```diff
@@ -0,0 +1,275 @@
+
+services:
+  trigger:
+    image: 'ghcr.io/triggerdotdev/trigger.dev:v4-beta'
```
Please pin to a static version rather than 'v4-beta' to prevent the template from breaking due to upstream changes.

ShadowArcanist commented on templates/compose/trigger-v4.yaml:
```diff
@@ -0,0 +1,275 @@
+      start_period: 120s
+
+  supervisor:
+    image: 'ghcr.io/triggerdotdev/supervisor:main'
```
Please pin to a static version rather than 'main' to prevent the template from breaking due to upstream changes.

ShadowArcanist commented on templates/compose/trigger-v4.yaml:
```diff
@@ -0,0 +1,275 @@
+      start_period: 30s
+
+  docker-proxy:
+    image: tecnativa/docker-socket-proxy:latest
```
Please pin to a static version rather than 'latest' to prevent the template from breaking due to upstream changes.

ShadowArcanist commented:
Instead of creating new compose file for v4, just updating the existing compose!

VictorCano committed:
feat(templates): replace Trigger.dev v3 with v4 (breaking change)

Replace the existing Trigger.dev v3 template with v4 instead of
maintaining both versions. This addresses maintainer feedback
to update the existing template rather than creating a new one.

Changes:
- Updated trigger.yaml with v4 architecture (8 services vs 6)
- Removed trigger-v4.yaml (merged into main template)
- Removed trigger-with-external-database.yaml (v4 requires ClickHouse/MinIO)
- Updated service-templates.json and service-templates-latest.json

Breaking Change: v4 is not backward compatible with v3
- v3 used coordinator + docker-provider
- v4 uses supervisor + docker-proxy + ClickHouse + MinIO
- New services: ClickHouse (analytics), MinIO (object storage)
- Existing v3 deployments will need manual migration

Tested: All 8 v4 containers deploy and run successfully

VictorCano commented:
Updated\! I've replaced the v3 template with v4 as requested instead of creating a separate file. All 8 containers tested and working.

ShadowArcanist commented:
> Updated! I've replaced the v3 template with v4 as requested instead of creating a separate file. All 8 containers tested and working.

Most of my comments were not addressed such us replacing the vulnerable minio image, restart policy, networks, service image versions, please check the comments again and implement the changes

eduhdev12 commented:
I've tried to replace the template, all containers are healthy, but I can't login, after requesting the magic link, everything blows up

also getting this response under the button `connect ECONNREFUSED 127.0.0.1:587`

ShadowArcanist commented:
> I've tried to replace the template, all containers are healthy, but I can't login, after requesting the magic link, everything blows up
> 
> also getting this response under the button `connect ECONNREFUSED 127.0.0.1:587`

Hey @VictorCano could you please check ^^

eduhdev12 commented:
Also, after setting up the email vars, the login is working. But the production tasks are not working at all, stuck in queue

VictorCano commented:
> > Updated! I've replaced the v3 template with v4 as requested instead of creating a separate file. All 8 containers tested and working.
> 
> Most of my comments were not addressed such us replacing the vulnerable minio image, restart policy, networks, service image versions, please check the comments again and implement the changes

I'm sorry. I had to migrate hosts for better testing capacity and accidentally pushed from the old one. I'm reviewing and will address the other issues as well.

VictorCano committed:
fix(trigger-v4): address PR review feedback from @ShadowArcanist

- Remove restart: unless-stopped from all services (Coolify adds these automatically)
- Remove network configurations from all services and networks section
- Pin MinIO image to secure version (CVE fix)
- Pin Trigger.dev image from v4-beta to v4.3.1
- Pin Supervisor image from main to v4.3.1
- Pin Docker Socket Proxy from latest to 0.4.2
- Pin ClickHouse from latest to 25.8-lts
- Pin PostgreSQL from 16-alpine to 16.6-alpine
- Pin Redis from 7 to 7.2.7

eduhdev12 commented:
@VictorCano The current update has some issues

`ghcr.io/triggerdotdev/trigger.dev:v4.3.1` -> `ghcr.io/triggerdotdev/trigger.dev:v4.3.0`
`tecnativa/docker-socket-proxy:0.4.2` -> `tecnativa/docker-socket-proxy:v0.4.2`
`clickhouse/clickhouse-server:25.8-lts` -> `clickhouse/clickhouse-server:25.8-alpine`
`ghcr.io/triggerdotdev/supervisor:v4.3.1` -> `ghcr.io/triggerdotdev/supervisor:v4.3.0`

Most of images have unknown manifest now

eduhdev12 commented:
Also minio is failing

```
dependency failed to start: container minio-q4gw80g8w48wsosgo00w4csg is unhealthy
```

peaklabs-dev commented on templates/compose/trigger.yaml:
```diff
@@ -1,146 +1,247 @@
-    environment:
-      SERVICE_URL_TRIGGER_3000: ""
-      <<: *common-env
+    image: 'ghcr.io/triggerdotdev/trigger.dev:v4.3.1'
```
Pin to an image version that exists

peaklabs-dev commented on templates/compose/trigger.yaml:
```diff
@@ -1,146 +1,247 @@
-      SERVICE_URL_TRIGGER_3000: ""
-      <<: *common-env
+    image: 'ghcr.io/triggerdotdev/trigger.dev:v4.3.1'
+    user: root
```
The container should if possible not need `root`

peaklabs-dev commented on templates/compose/trigger.yaml:
```diff
@@ -1,146 +1,247 @@
-# documentation: https://trigger.dev
-# slogan: The open source Background Jobs framework for TypeScript
+# documentation: https://trigger.dev/docs/v4/open-source-self-hosting
```
Not sure what is going on with their docs but this link does not work

peaklabs-dev commented on templates/compose/trigger.yaml:
```diff
@@ -1,146 +1,247 @@
-      SECURE_CONNECTION: "false"
-      PLATFORM_SECRET: $SERVICE_PASSWORD_64_COORDINATOR
-      HTTP_SERVER_PORT: 9020
+      - CLICKHOUSE_USER=default
```
Autogenerate the user with the Coolify magic

peaklabs-dev commented on templates/compose/trigger.yaml:
```diff
@@ -1,146 +1,247 @@
+      start_period: 30s
+
+  minio:
+    image: ghcr.io/coollabsio/minio:RELEASE.2025-10-15T17-29-55Z
```
Do we really need minio or can we use external s3 as well because it is no longer maintained?

VictorCano replied:
> Do we really need minio or can we use external s3 as well because it is no longer maintained?

Minio is not needed, It's just there for ease of install but any S3 compatible can be used. I followed their [Docker Compose](https://github.com/triggerdotdev/trigger.dev/blob/main/hosting/docker/webapp/docker-compose.yml)

peaklabs-dev replied:
I removed it as it is in maintenance only mode.

peaklabs-dev commented on templates/compose/trigger.yaml:
```diff
@@ -1,146 +1,247 @@
+      retries: 5
+      start_period: 10s
+
+volumes:
```
This will be added automatically so this can be removed

peaklabs-dev commented on templates/compose/trigger.yaml:
```diff
@@ -1,146 +1,247 @@
+      start_period: 10s
+
+  electric:
+    image: electricsql/electric:1.0.24
```
There is a much newer version available use that

peaklabs-dev commented on templates/compose/trigger.yaml:
```diff
@@ -1,146 +1,247 @@
+      start_period: 5s
+
   postgresql:
     image: postgres:16-alpine
```
This can probably be changed to 17 or 18

VictorCano replied:
I agree, but changing it from 16 would require further testing. [Their docker-compose.yml](https://github.com/triggerdotdev/trigger.dev/blob/main/hosting/docker/webapp/docker-compose.yml) has version 14 as default, but 16 is what we're running and had no issues.

peaklabs-dev replied:
Sounds good then we will leave it.

peaklabs-dev commented on templates/compose/trigger.yaml:
```diff
@@ -1,146 +1,247 @@
-    environment:
-      <<: *common-env
+  supervisor:
+    image: 'ghcr.io/triggerdotdev/supervisor:v4.3.1'
```
This is also an invalid version

peaklabs-dev commented on templates/compose/trigger.yaml:
```diff
@@ -1,146 +1,247 @@
-      <<: *common-env
+  supervisor:
+    image: 'ghcr.io/triggerdotdev/supervisor:v4.3.1'
+    user: root
```
The container should if possible not need root

peaklabs-dev commented on templates/compose/trigger.yaml:
```diff
@@ -1,146 +1,247 @@
+      start_period: 10s
+
+  clickhouse:
+    image: clickhouse/clickhouse-server:25.8-lts
```
Invalid tag

peaklabs-dev commented on templates/compose/trigger.yaml:
```diff
@@ -1,146 +1,247 @@
-  redis:
-    image: "redis:7"
+  docker-proxy:
+    image: tecnativa/docker-socket-proxy:0.4.2
```
Invalid tag

VictorCano committed:
fix(trigger-v4): resolve deployment and authentication issues

Addresses remaining issues discovered during testing of the Trigger.dev v4 template
after initial PR review feedback implementation.

Image Version Fixes:
- Downgrade trigger and supervisor images from v4.3.1 to v4.3.0 (v4.3.1 manifest not available)
- Fix ClickHouse image tag from 25.8-lts to 25.8 (correct tag format)
- Add 'v' prefix to docker-proxy image tag for consistency (v0.4.2)

Email/Authentication Fixes:
- Change EMAIL_TRANSPORT default from 'smtp' to 'resend' (aligns with provided RESEND_API_KEY)
- Add FROM_EMAIL default 'onboarding@resend.dev' for local testing
- Fix: Resend API requires FROM address to send magic links

MinIO Healthcheck Fix:
- Replace curl-based healthcheck with 'mc ready local' command
- Resolves container health issues (curl not available in ghcr.io/coollabsio/minio image)

Environment Variable Cleanup:
- Remove unnecessary REMIX_APP_PORT environment variable
- Remove PORT variable (not needed for Trigger.dev v4)

Docker Compose Cleanup:
- Remove volumes section (Coolify creates named volumes automatically)

These changes resolve the "domain is invalid" error during magic link authentication
and ensure all services start correctly with proper health checks.

Fixes remaining issues from PR #7808 review by @ShadowArcanist

VictorCano commented:
I'm addressing all issues, will make the remaining changes from @peaklabs-dev tonight and test in different envs for review. I believe in 12h max all should be addressed.

VictorCano committed:
fix(trigger-v4): address remaining PR review feedback

- Update documentation link to self-hosting overview page
- Add comment explaining root user requirement for bootstrap
- Replace hardcoded usernames with Coolify magic variables:
  - MinIO: admin → ${SERVICE_USER_MINIO}
  - ClickHouse: default → ${SERVICE_USER_CLICKHOUSE}

This improves security by using auto-generated credentials instead of hardcoded values.

References:
- PR #7808 review by peaklabs-dev
- Trigger.dev upstream docker-compose.yml for root user justification

VictorCano commented:
I believe all major changes were made and tested, they should also fix the queue issues. If possible I'd like some advice on the open conversations.

Once we're in a final version I'll push the docs in the other fork. As well including details on Auth methods, root user on Trigger.dev web app details, and also S3/Minio.

Thanks @ShadowArcanist  and @peaklabs-dev  for your feedback, we're increasing our usage on Coolify and we've encountered some templates that we need to make our own. Now knowing better of your support I'll tell my team to open PRs for issues we encounter, as they're more experienced and have more time to support.

peaklabs-dev committed:
Merge branch 'next' into feat/trigger-dev-v4-template

peaklabs-dev committed:
feat(service): improve trigger v4

- remove most start_periods as it just slows deployments down
- remove minio as it is in maintenance mode
- autogenerate ENCRYPTION_KEY
- remove empty defaults from many envs
- make some more envs configurable
- updated electricsql image tag
- fix formatting, naming and order

peaklabs-dev committed:
chore(service): remove unused logo file

peaklabs-dev reviewed (approved):
Thanks for the PR and the fixes. I improved quite a few more things here https://github.com/coollabsio/coolify/pull/7808/changes/a6d5f5ab4edc1c28b42cb55e8b40b005e85d6a51. 

For next time please do not commit the service json files to the PR.

VictorCano commented:
@peaklabs-dev Thanks for letting me know, sorry about that.

I'll update the docs as soon as I can. I have a quick work trip from tomorrow until sunday, will try to work on it when I can.

Thanks again for you and Coolify team for your support and for a great product!

Cheers!

ShadowArcanist commented:
> @peaklabs-dev Thanks for letting me know, sorry about that.
> 
> I'll update the docs as soon as I can. I have a quick work trip from tomorrow until sunday, will try to work on it when I can.
> 
> Thanks again for you and Coolify team for your support and for a great product!
> 
> Cheers!

Ping me on discord (on docs contributors channel) after you sent PR to our docs repo so I can check it out!

Safe travels mate!

eduhdev12 commented:
@ShadowArcanist @VictorCano Tested, dev is working, deploying is working. Running tasks in production env is not working, the error from Supervisor container is

```
"json":{"message":"network trigger-net not found"}},"containerId":"8bb24c6fb1b755376bb395fab0e3b67bef6f75881fd5bf02f6e7229a1f97876e"}
```

_So `trigger-net` is required_

nvm, `DOCKER_RUNNER_NETWORKS` should be default to `coolify`, the shared network

ShadowArcanist commented:
> @ShadowArcanist @VictorCano Tested, dev is working, deploying is working. Running tasks in production env is not working, the error from Supervisor container is
> 
> ```
> "json":{"message":"network trigger-net not found"}},"containerId":"8bb24c6fb1b755376bb395fab0e3b67bef6f75881fd5bf02f6e7229a1f97876e"}
> ```
> 
> _So `trigger-net` is required_
> 
> nvm, `DOCKER_RUNNER_NETWORKS` should be default to `coolify`, the shared network

compose based apps are not added to the default coolify network by default (unless you enable "connect to predefined network" option), coolify will create a docker network with the uuid of the service as the name and put all services in the compose on the network and adds coolify-proxy to this network.
